Switch to using JSON output of `rclone size`

kirelagin opened the pull request:
Example output from `rclone size --json`:

```text
{"count":1,"bytes":215935}
```

Much easier to parse, even with grep.

Obsoletes #42.

kirelagin committed:
Switch to using JSON output of `rclone size`

Previous parsing of `rclone size` output with `grep` was pretty
unreliable, since the output format was changing all the time. On the
other hand, the JSON output seems to be much more stable, so use it
instead.

davidmcote commented:
Thanks for creating this patch kirelagin!  I'm able to run `git annex testremote --fast` successfully with `rclone v1.57.0`.

kirelagin commented:
Hi @DanielDent, would it be possible to merge this?

zarelit commented on git-annex-remote-rclone:
```diff
@@ -198,24 +198,20 @@ while read -r line; do
-					echo CHECKPRESENT-UNKNOWN "$key" "remote currently unavailable or git-annex-remote-rclone failed to parse rclone output"
+					echo CHECKPRESENT-FAILURE "$key"
 				fi
+			else
```
@kirelagin When the directory is non existent `check_result` value is `Failed to size: directory not found`. This will make the remote return `CHECKPRESENT-UNKNOWN` instead of `CHECKPRESENT-FAILURE`

zarelit replied:
Also I think this is an upstream issue because we asked for json output and instead we get an empty output and a human-readable stderr, which is kind the opposite of the idea of a `--json` flag usage

kirelagin replied:
Hmm, I can’t reproduce this or I don’t understand the situation you are describing. What do you mean by “the directory is non existent”? Here is what I am getting:

```
› rclone size --json b2:kirelagin-annex/foo/bar/baz
{"count":0,"bytes":0}
```

zarelit replied:
@kirelagin can you try with a backend that is not an object storage? This is rclone being inconsistent in the output

```
$ rclone size --json zarel-blaze:test-bucket-annex/yay/yayy
{"count":0,"bytes":0}

$ rclone size --json zarel-box:thisfolderdoesnotexist
2022/04/20 20:51:16 ERROR : : error listing: directory not found
2022/04/20 20:51:16 Failed to size with 2 errors: last error was: directory not found

$ rclone size --json zarel-drive:test/folder/file
2022/04/20 20:51:35 Failed to size: directory not found

```

rclone version is 1.57.0

kirelagin replied:
Yes, you are right, I am getting this with a non-object backend too. I am not sure what the guarantees of the JSON output are, but, I guess, we can assume that if the exit code is >0, then we are not getting JSON.

kirelagin committed:
Handle the case when the output is not JSON

With non-object-storage backends, rclone can return a plaintext error
even when called with --json.

We handle this case by checking for a non-zero exit code and falling
back to the old parsing logic.

DanielDent commented:
At this point, it's been long enough that I'm fine with requiring 1.53.3 or even newer frankly. Users can easily use an older version of `git-annex-remote-rclone` or a newer version of `rclone` if they need to (both are nearly dependency free), and Debian users have a compatible version of `git-annex-remote-rclone` packaged anyway.

Actually parsing the JSON is probably better than playing games with regexes, though it may introduce extra dependency requirements users.

yarikoptic commented:
@kirelagin would you have interest&time to update this PR to go against current `master` ?  or should one of us to take over? (it must be so much fun to parse json in bash!)

jleeuwes commented:
This change is included in #57 which was based on current `master`.

yarikoptic commented:
doh, right! So we can close this one!
